src/executor/plan_executor.cpp

```diff
@@ -105,38 +52,110 @@ void PlanExecutor::ExecutePlan(
   plan->GetOutputColumns(columns);
   codegen::BufferingConsumer consumer{columns, context};
 
-  // Prepare parameter: TODO Combine with executor context when legacy is gone
-  codegen::QueryParameters parameters{*plan, params};
+  std::unique_ptr<executor::ExecutorContext> executor_context(
+      new executor::ExecutorContext(txn,
+                                    codegen::QueryParameters(*plan, params)));
 
-  // Compile and execute the query
+  // Compile the query
   codegen::Query *query = codegen::QueryCache::Instance().Find(plan);
   if (query == nullptr) {
     codegen::QueryCompiler compiler;
-    auto compiled_query =
-        compiler.Compile(*plan, parameters.GetQueryParametersMap(), consumer);
-
-    compiled_query->Execute(*executor_context.get(), parameters,
-                            reinterpret_cast<char *>(consumer.GetState()));
+    auto compiled_query = compiler.Compile(
+        *plan, executor_context->GetParams().GetQueryParametersMap(), consumer);
+    query = compiled_query.get();
     codegen::QueryCache::Instance().Add(plan, std::move(compiled_query));
-  } else {
-    query->Execute(*executor_context.get(), parameters,
-                   reinterpret_cast<char *>(consumer.GetState()));
   }
 
-  // Iterate over results
-  const auto &results = consumer.GetOutputTuples();
-  for (const auto &tuple : results) {
-    for (uint32_t i = 0; i < tuple.tuple_.size(); i++) {
-      auto column_val = tuple.GetValue(i);
-      auto str = column_val.IsNull() ? "" : column_val.ToString();
-      LOG_TRACE("column content: [%s]", str.c_str());
-      result.push_back(std::move(str));
+  query->Execute(std::move(executor_context), consumer,
+                 [&](executor::ExecuteResult execute_result) {
```

Comment:
Try to be explicit in your capture list to make it clear what's needed in your function.